src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Versioning/ControlPlaneCollectionUpgradeManager.cs

```diff
@@ -0,0 +1,79 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Azure.Documents;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.CosmosDb.Configs;
+using Microsoft.Health.CosmosDb.Features.Storage;
+using Microsoft.Health.CosmosDb.Features.Storage.Versioning;
+
+namespace Microsoft.Health.ControlPlane.CosmosDb.Features.Storage.Versioning
+{
+    public class ControlPlaneCollectionUpgradeManager : IUpgradeManager
+    {
+        private readonly IEnumerable<ICollectionUpdater> _collectionUpdater;
+        private readonly CosmosDataStoreConfiguration _configuration;
+        private readonly CosmosCollectionConfiguration _collectionConfiguration;
+        private readonly ICosmosDbDistributedLockFactory _lockFactory;
+        private readonly ILogger<ControlPlaneCollectionUpgradeManager> _logger;
+
+        /// <summary>
+        /// This integer should be incremented when changing any value in the
+        /// UpdateIndexAsync function
+        /// </summary>
+        internal const int CollectionSettingsVersion = 1;
+
+        public ControlPlaneCollectionUpgradeManager(
+            IEnumerable<IControlPlaneCollectionUpdater> collectionUpdater,
+            CosmosDataStoreConfiguration configuration,
+            ICosmosDbDistributedLockFactory lockFactory,
+            IOptionsMonitor<CosmosCollectionConfiguration> namedCosmosCollectionConfigurationAccessor,
+            ILogger<ControlPlaneCollectionUpgradeManager> logger)
+        {
+            EnsureArg.IsNotNull(collectionUpdater, nameof(collectionUpdater));
+            EnsureArg.IsNotNull(configuration, nameof(configuration));
+            EnsureArg.IsNotNull(lockFactory, nameof(lockFactory));
+            EnsureArg.IsNotNull(namedCosmosCollectionConfigurationAccessor, nameof(namedCosmosCollectionConfigurationAccessor));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _collectionUpdater = collectionUpdater;
+            _configuration = configuration;
+            _lockFactory = lockFactory;
+            _collectionConfiguration = namedCosmosCollectionConfigurationAccessor.Get(Constants.CollectionConfigurationName);
+            _logger = logger;
+        }
+
+        public async Task SetupCollectionAsync(IDocumentClient documentClient, DocumentCollection collection)
```

Comment:
I think you can create a base class and move this method there then have this and FhirCollectionUpgradeManager both derived from that class.

Reply:
refactored